// vue3-snackbar/dist/vue3-snackbar.js
import { h as k, computed as l, ref as O, onMounted as T, watch as H, isVNode as z, openBlock as h, createElementBlock as S, normalizeClass as B, unref as w, normalizeStyle as R, renderSlot as C, createElementVNode as M, createVNode as A, mergeProps as W, createCommentVNode as _, toDisplayString as I, inject as X, nextTick as Y, getCurrentScope as ee, onScopeDispose as se, getCurrentInstance as Z, onUnmounted as te, createBlock as j, Teleport as ae, TransitionGroup as oe, withCtx as D, Fragment as ne, renderList as P, createSlots as re } from "vue";
var ie = "M8.27,3L3,8.27V15.73L8.27,21H15.73C17.5,19.24 21,15.73 21,15.73V8.27L15.73,3M9.1,5H14.9L19,9.1V14.9L14.9,19H9.1L5,14.9V9.1M11,15H13V17H11V15M11,7H13V13H11V7", le = "M12,2L1,21H23M12,6L19.53,19H4.47M11,10V14H13V10M11,16V18H13V16", ue = "M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z", ce = "M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z", de = "M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z";
const fe = {
  transform: "rotate(var(--r, 0deg)) scale(var(--sx, 1), var(--sy, 1))"
}, q = {
  fill: "currentColor"
}, F = {
  mdi: {
    size: 24,
    viewbox: "0 0 24 24"
  },
  "simple-icons": {
    size: 24,
    viewbox: "0 0 24 24"
  },
  default: {
    size: 0,
    viewbox: "0 0 0 0"
  }
}, G = {
  name: "icon",
  props: {
    type: {
      type: String,
      default: "mdi"
    },
    faIcon: {
      type: Object,
      default: null
    },
    path: {
      type: [String, Object, Array]
    },
    size: {
      type: [Number, String],
      default: 24
    },
    viewbox: String,
    flip: {
      type: String,
      validator: (s) => ["horizontal", "vertical", "both"].includes(s)
    },
    rotate: {
      type: [Number, String],
      default: 0
    }
  },
  setup(s) {
    if (!s.path && !s.faIcon)
      return console.warn("vue3-icon requires either a 'path' or an 'fa-icon' property"), () => k("div");
    const a = l(() => {
      var d;
      return ((d = s.faIcon) == null ? void 0 : d.prefix) || s.type;
    }), o = l(() => parseInt(s.rotate, 10)), e = l(() => F[a.value] || F.default), r = l(() => parseInt(s.size, 10) || e.value.size), i = l(() => s.faIcon ? `0 0 ${s.faIcon.icon[0]} ${s.faIcon.icon[1]}` : !1), u = l(() => i.value || s.viewbox || e.value.viewbox), m = l(() => ({
      ...fe,
      "--sx": ["both", "horizontal"].includes(s.flip) ? "-1" : "1",
      "--sy": ["both", "vertical"].includes(s.flip) ? "-1" : "1",
      "--r": isNaN(o.value) ? o.value : o.value + "deg"
    })), c = l(() => {
      var d;
      return s.faIcon ? (d = s.faIcon) == null ? void 0 : d.icon[4] : s.type === "simple-icons" && typeof s.path == "object" ? s.path.path : s.path;
    }), g = l(() => a.value === "fad" ? (console.warn("vue3-icon does not currently support Duotone FontAwesome icons"), k("path")) : Array.isArray(s.path) ? k(
      "g",
      { style: { ...q } },
      s.path.map((d) => typeof d == "string" ? k("path", { d }) : k("path", { ...d }))
    ) : k("path", { d: c.value, style: { ...q } }));
    return () => k(
      "svg",
      {
        style: m.value,
        class: ["vue3-icon"],
        width: r.value,
        height: r.value,
        viewBox: u.value
      },
      [g.value]
    );
  }
}, me = ["role"], ge = { class: "vue3-snackbar-message-wrapper" }, pe = {
  key: 0,
  class: "vue3-snackbar-message-icon"
}, ve = { class: "vue3-snackbar-message-content" }, be = {
  key: 0,
  class: "vue3-snackbar-message-badge"
}, ye = { class: "vue3-snackbar-message-title" }, he = {
  key: 0,
  class: "vue3-snackbar-message-additional"
}, ke = /* @__PURE__ */ M("div", { class: "spacer" }, null, -1), Se = { class: "vue3-snackbar-message-close" }, Ce = {
  __name: "Vue3SnackbarMessage",
  props: {
    borderClass: {
      type: String,
      default: ""
    },
    message: {
      type: Object,
      default: () => ({})
    },
    messageClass: {
      type: String,
      default: ""
    },
    messageActionClass: {
      type: String,
      default: "vue3-snackbar-message-action"
    },
    dense: {
      type: Boolean,
      default: !1
    },
    iconPresets: {
      type: Object,
      required: !0
    }
  },
  emits: ["dismiss", "click:action"],
  setup(s, { emit: a }) {
    const o = a, e = s;
    let r = null, i = null, u = O(!1);
    const m = () => {
      const t = !e.message.duration && !e.message.dismissible ? 4e3 : e.message.duration;
      r = setTimeout(g, t);
    };
    T(() => {
      m();
    }), H(
      () => e.message.count,
      (t) => {
        if (t === 1) return !1;
        clearTimeout(r), clearTimeout(i), i = setTimeout(() => {
          u.value = !1;
        }, 1e3), u.value = !0, m();
      }
    );
    const c = () => {
      r && clearTimeout(r), g();
    }, g = () => {
      o("dismiss", e.message);
    }, d = l(
      () => Object.assign(
        {},
        {
          success: {
            path: ue
          },
          info: {
            path: de
          },
          warning: {
            path: le
          },
          error: {
            path: ie
          }
        },
        e.iconPresets
      )
    ), p = l(() => {
      const t = d.value[e.message.type];
      return e.message.type && !t && console.warn("[Vue3 Snackbar] Couldn't find icon preset for type " + e.message.type), t ? (t.type = "mdi", t) : e.message.icon && typeof e.message.icon == "object" ? e.message.icon : {
        path: "",
        type: "default"
      };
    }), v = l(() => {
      var t;
      return e.message.type === "error" ? "alert" : ((t = e.message) == null ? void 0 : t.role) ?? "status";
    }), n = l(() => {
      const t = e.message.action;
      if (t) {
        if (z(t)) return t;
        if (typeof t == "object" && (t.render || t.setup))
          return t;
        if (typeof t == "function") {
          const f = t();
          return z(f) ? t : t();
        } else return null;
      } else return null;
    });
    return (t, f) => (h(), S("article", {
      class: B(["vue3-snackbar-message", [
        e.message.type || "custom",
        e.messageClass,
        e.borderClass,
        {
          "has-background": e.message.background,
          "has-border": e.borderClass,
          "is-dense": e.dense,
          "shake-baby-shake": w(u)
        }
      ]]),
      role: v.value,
      style: R({
        "--message-background": e.message.background,
        "--message-text-color": e.message.textColor,
        "--message-icon-color": e.message.iconColor
      })
    }, [
      C(t.$slots, "message-inner", {
        message: e.message
      }, () => [
        M("div", ge, [
          p.value ? (h(), S("div", pe, [
            C(t.$slots, "message-icon", {
              message: e.message,
              icon: p.value
            }, () => [
              A(w(G), W(p.value, { role: "img" }), null, 16)
            ])
          ])) : _("", !0),
          M("div", ve, [
            C(t.$slots, "message-badge", {
              message: e.message,
              count: e.message.count
            }, () => [
              e.message.count > 1 ? (h(), S("div", be, I(e.message.count), 1)) : _("", !0)
            ]),
            C(t.$slots, "message-content", {
              message: e.message,
              title: e.message.title,
              text: e.message.text
            }, () => [
              M("div", ye, I(e.message.title || e.message.text), 1),
              e.message.title && e.message.text ? (h(), S("div", he, I(e.message.text), 1)) : _("", !0)
            ]),
            t.$slots["message-action"] ? (h(), S("div", {
              key: 0,
              class: B(e.messageActionClass),
              onClick: f[0] || (f[0] = (L) => o("click:action", { message: e.message }))
            }, [
              C(t.$slots, "message-action", {
                message: e.message,
                isDismissible: e.message.dismissible,
                dismiss: c
              }, () => [
                A(w(n), {
                  message: e.message,
                  "is-dismissible": e.message.dismissible,
                  dismiss: c
                }, null, 8, ["message", "is-dismissible"])
              ])
            ], 2)) : _("", !0)
          ]),
          ke,
          M("div", Se, [
            C(t.$slots, "message-close-icon", {
              message: e.message,
              isDimissible: e.message.dismissible,
              isDismissible: e.message.dismissible,
              dismiss: c
            }, () => [
              e.message.dismissible !== !1 ? (h(), S("button", {
                key: 0,
                onClick: c
              }, [
                A(w(G), {
                  type: "mdi",
                  path: w(ce)
                }, null, 8, ["path"])
              ])) : _("", !0)
            ])
          ])
        ])
      ])
    ], 14, me));
  }
}, we = typeof window < "u" ? HTMLElement : Object, Me = {
  /* ******************************************
   * LOCATION PROPS
   ****************************************** */
  top: {
    type: Boolean,
    default: !1
  },
  bottom: {
    type: Boolean,
    default: !1
  },
  left: {
    type: Boolean,
    default: !1
  },
  right: {
    type: Boolean,
    default: !1
  },
  /* ******************************************
   * COLOUR PROPS
   ****************************************** */
  success: {
    type: String,
    default: "#4caf50"
  },
  error: {
    type: String,
    default: "#ff5252"
  },
  warning: {
    type: String,
    default: "#fb8c00"
  },
  info: {
    type: String,
    default: "#2196f3"
  },
  iconPresets: {
    type: Object,
    default: null
  },
  messageTextColor: {
    type: String,
    default: "#fff"
  },
  messageIconColor: {
    type: String,
    default: "currentColor"
  },
  /* ******************************************
   * OTHER PROPS
   ****************************************** */
  attach: {
    type: [String, we],
    default: "body"
  },
  border: {
    type: String,
    default: "",
    validator: (s) => ["top", "bottom", "left", "right", ""].includes(s)
  },
  backgroundOpacity: {
    type: [String, Number],
    default: 0.12,
    validator: (s) => !isNaN(parseFloat(s)) && isFinite(s)
  },
  backgroundColor: {
    type: String,
    default: "currentColor"
  },
  baseBackgroundColor: {
    type: String,
    default: "#fff"
  },
  duration: {
    type: [Number, String],
    default: null
  },
  messageClass: {
    type: String
  },
  messageActionClass: {
    type: String,
    default: "vue3-snackbar-message-action"
  },
  zIndex: {
    type: Number,
    default: 1e4
  },
  dense: {
    type: Boolean,
    default: !1
  },
  reverse: {
    type: Boolean,
    default: !1
  },
  limit: {
    type: Number,
    default: null
  },
  groups: {
    type: Boolean,
    default: !1
  },
  shadow: {
    type: Boolean,
    default: !1
  },
  dismissOnActionClick: {
    type: Boolean,
    default: !0
  }
};
function $e(s) {
  return s && s.__esModule && Object.prototype.hasOwnProperty.call(s, "default") ? s.default : s;
}
var N = { exports: {} };
function E() {
}
E.prototype = {
  on: function(s, a, o) {
    var e = this.e || (this.e = {});
    return (e[s] || (e[s] = [])).push({
      fn: a,
      ctx: o
    }), this;
  },
  once: function(s, a, o) {
    var e = this;
    function r() {
      e.off(s, r), a.apply(o, arguments);
    }
    return r._ = a, this.on(s, r, o);
  },
  emit: function(s) {
    var a = [].slice.call(arguments, 1), o = ((this.e || (this.e = {}))[s] || []).slice(), e = 0, r = o.length;
    for (e; e < r; e++)
      o[e].fn.apply(o[e].ctx, a);
    return this;
  },
  off: function(s, a) {
    var o = this.e || (this.e = {}), e = o[s], r = [];
    if (e && a)
      for (var i = 0, u = e.length; i < u; i++)
        e[i].fn !== a && e[i].fn._ !== a && r.push(e[i]);
    return r.length ? o[s] = r : delete o[s], this;
  }
};
N.exports = E;
N.exports.TinyEmitter = E;
var Le = N.exports, _e = Le, xe = new _e();
const V = /* @__PURE__ */ $e(xe), $ = {
  $on: (...s) => V.on(...s),
  $once: (...s) => V.once(...s),
  $off: (...s) => V.off(...s),
  $emit: (...s) => V.emit(...s)
}, y = O([]), U = Symbol();
function Pe() {
  const s = X(U);
  if (!s)
    throw new Error("No Snackbar provided!");
  return s;
}
const qe = {
  install: (s, a = {}) => {
    const { disableGlobals: o = !1 } = a, e = {
      add: (r) => {
        $.$emit("add", r);
      },
      clear: () => {
        $.$emit("clear");
      }
    };
    o !== !0 && (s.config.globalProperties.$snackbar = e, typeof window < "u" && (window.$snackbar = e)), s.provide(U, e);
  }
};
function Ve(s) {
  return ee() ? (se(s), !0) : !1;
}
function J(s) {
  return typeof s == "function" ? s() : w(s);
}
const K = typeof window < "u" && typeof document < "u";
typeof WorkerGlobalScope < "u" && globalThis instanceof WorkerGlobalScope;
const Ae = (s) => s != null;
function Oe(s) {
  return Z();
}
function Ie(s, a = !0, o) {
  Oe() ? T(s, o) : a ? s() : Y(s);
}
function Be(s) {
  var a;
  const o = J(s);
  return (a = o == null ? void 0 : o.$el) != null ? a : o;
}
const Te = K ? window : void 0, He = K ? window.document : void 0;
function Ne() {
  const s = O(!1), a = Z();
  return a && T(() => {
    s.value = !0;
  }, a), s;
}
function Ee(s) {
  const a = Ne();
  return l(() => (a.value, !!s()));
}
function ze(s, a, o = {}) {
  const { window: e = Te, ...r } = o;
  let i;
  const u = Ee(() => e && "MutationObserver" in e), m = () => {
    i && (i.disconnect(), i = void 0);
  }, c = l(() => {
    const v = J(s), n = (Array.isArray(v) ? v : [v]).map(Be).filter(Ae);
    return new Set(n);
  }), g = H(
    () => c.value,
    (v) => {
      m(), u.value && v.size && (i = new MutationObserver(a), v.forEach((n) => i.observe(n, r)));
    },
    { immediate: !0, flush: "post" }
  ), d = () => i == null ? void 0 : i.takeRecords(), p = () => {
    m(), g();
  };
  return Ve(p), {
    isSupported: u,
    stop: p,
    takeRecords: d
  };
}
function je(s = {}) {
  const {
    document: a = He,
    selector: o = "html",
    observe: e = !1,
    initialValue: r = "ltr"
  } = s;
  function i() {
    var m, c;
    return (c = (m = a == null ? void 0 : a.querySelector(o)) == null ? void 0 : m.getAttribute("dir")) != null ? c : r;
  }
  const u = O(i());
  return Ie(() => u.value = i()), e && a && ze(
    a.querySelector(o),
    () => u.value = i(),
    { attributes: !0 }
  ), l({
    get() {
      return u.value;
    },
    set(m) {
      var c, g;
      u.value = m, a && (u.value ? (c = a.querySelector(o)) == null || c.setAttribute("dir", u.value) : (g = a.querySelector(o)) == null || g.removeAttribute("dir"));
    }
  });
}
const Fe = {
  __name: "Vue3Snackbar",
  props: { ...Me },
  emits: ["added", "dismissed", "removed", "cleared", "click:action"],
  setup(s, { emit: a }) {
    const o = je(), e = s, r = a, i = l(() => ({
      "is-top": e.top,
      "is-bottom": e.top === !1 && e.bottom,
      "is-left": e.left,
      "is-right": e.left === !1 && e.right,
      "is-middle": e.top === !1 && e.bottom === !1,
      "is-centre": e.left === !1 && e.right === !1,
      "has-shadow": e.shadow,
      "is-rtl": o.value === "rtl"
    })), u = l(() => ({
      "--success-colour": e.success,
      "--error-colour": e.error,
      "--warning-colour": e.warning,
      "--info-colour": e.info,
      "--snackbar-zindex": e.zIndex,
      "--background-opacity": e.backgroundOpacity,
      "--background-color": e.backgroundColor,
      "--base-background-color": e.baseBackgroundColor,
      "--message-text-color": e.messageTextColor,
      "--message-icon-color": e.messageIconColor
    })), m = l(() => e.border ? `border-${e.border}` : ""), c = (n) => Math.abs(n.split("").reduce((t, f) => (t << 5) - t + f.charCodeAt(0) | 0, 0));
    let g = 1;
    $.$on("add", (n) => {
      r("added", n), n.group || (n.group = c(`${n.type}${n.title}${n.text}`).toString(16)), e.duration && !n.duration && n.duration !== 0 && (n.duration = +e.duration);
      const t = n.group && y.value.find((f) => f.group === n.group);
      if (e.groups === !1 || !t) {
        const f = {
          ...n,
          id: g,
          count: 1
        };
        y.value.push(f), g++;
      } else
        t.count++;
    });
    const d = l(() => {
      const n = e.limit ? y.value.slice(e.limit * -1) : [...y.value];
      return e.reverse ? n.reverse() : n;
    });
    H(d, (n) => {
      const t = n.map((b) => b.id), L = y.value.map((b) => b.id).filter((b) => !t.includes(b));
      for (const b of L) {
        const x = y.value.find((Q) => Q.id === b);
        x && p(x, !1);
      }
    }), $.$on("clear", () => {
      r("cleared"), y.value = [];
    }), te(() => {
      $.$off("add"), $.$off("clear");
    });
    const p = (n, t = !1) => {
      r(t ? "dismissed" : "removed", n), y.value = y.value.filter((f) => f.id !== n.id);
    }, v = (n) => {
      r("click:action", n), e.dismissOnActionClick && p(n.message, !0);
    };
    return (n, t) => (h(), j(ae, {
      to: e.attach
    }, [
      M("section", {
        id: "vue3-snackbar--container",
        class: B([[i.value], "vue3-snackbar"]),
        style: R(u.value),
        "aria-live": "polite"
      }, [
        A(oe, {
          name: "vue3-snackbar-message",
          tag: "div",
          class: "vue3-snackbar--messages"
        }, {
          default: D(() => [
            (h(!0), S(ne, null, P(d.value, (f) => (h(), j(Ce, {
              key: f.id,
              message: f,
              "message-class": e.messageClass,
              "message-action-class": e.messageActionClass,
              dense: e.dense,
              "border-class": m.value,
              "icon-presets": e.iconPresets,
              onDismiss: t[0] || (t[0] = (L) => p(L, !0)),
              "onClick:action": v
            }, re({ _: 2 }, [
              P(n.$slots, (L, b) => ({
                name: b,
                fn: D((x) => [
                  C(n.$slots, b, W({ ref_for: !0 }, x))
                ])
              }))
            ]), 1032, ["message", "message-class", "message-action-class", "dense", "border-class", "icon-presets"]))), 128))
          ]),
          _: 3
        })
      ], 6)
    ], 8, ["to"]));
  }
};
export {
  y as SnackbarMessages,
  qe as SnackbarService,
  Fe as Vue3Snackbar,
  Pe as useSnackbar
};
